File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int solve(vector<int>&arr,int n,int i,vector<int>&dp){
        if(i>=n)return 0;
        if(dp[i]!=-1)return dp[i];
        int take=arr[i]+solve(arr,n,i+2,dp);
        int notake=0+solve(arr,n,i+1,dp);
        return dp[i]=max(take,notake);
    }
    int rob(vector<int>& nums) {
        int n=nums.size();
        vector<int>dp(n+1,-1);
        return solve(nums,n,0,dp);
    }
};

File: 0198-house-robber/0198-house-robber.java
class Solution {


    int solve(int i,int[]nums,int[]dp){
        if(i==0)return nums[i];
        if(i<0)return 0;

        if(dp[i]!=-1){
            return dp[i];
        }

        int pick=nums[i]+solve(i-2,nums,dp);
        int nonpick=0+solve(i-1,nums,dp);

        dp[i]=Math.max(pick,nonpick);

        return dp[i];
    }


    public int rob(int[] nums) {
        int []dp=new int[nums.length];
        Arrays.fill(dp,-1);

        int result=solve(nums.length-1,nums,dp);
        return result;

        
    }
}